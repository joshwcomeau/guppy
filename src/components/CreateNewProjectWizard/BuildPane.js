// @flow
import React, { PureComponent } from 'react';
import styled from 'styled-components';
import IconBase from 'react-icons-kit';
import { check } from 'react-icons-kit/feather/check';

import { COLORS } from '../../constants';
import createProject from '../../services/create-project.service';

import Spacer from '../Spacer';
import WhimsicalInstaller from '../WhimsicalInstaller';
import BuildStepProgress from './BuildStepProgress';

import type { BuildStep, Status } from './types';
import type { ProjectType, ProjectInternal } from '../../types';

const BUILD_STEPS = {
  installingCliTool: {
    copy: 'Installing build tool',
  },
  creatingProjectDirectory: {
    copy: 'Creating project directory',
  },
  installingDependencies: {
    copy: 'Installing dependencies',
    additionalCopy: 'This step can take a while...',
  },
  guppification: {
    copy: 'Persisting project info',
  },
};

const BUILD_STEP_KEYS: Array<BuildStep> = Object.keys(BUILD_STEPS);

type Props = {
  projectName: string,
  projectType: ?ProjectType,
  projectIcon: ?string,
  status: Status,
  projectHomePath: string,
  handleCompleteBuild: (project: ProjectInternal) => void,
};

type State = {
  currentBuildStep: BuildStep,
  isCompleted: boolean,
  runInstaller: boolean,
};

class BuildPane extends PureComponent<Props, State> {
  timeoutId: ?number;
  state = {
    currentBuildStep: BUILD_STEPS[0],
    isCompleted: false,
    runInstaller: false,
  };

  componentDidUpdate(prevProps: Props) {
    if (
      prevProps.status === 'filling-in-form' &&
      this.props.status === 'building-project'
    ) {
      // We want to wait a bit before starting the whimsy animation, because
      // at this very moment, the pane is just beginning to unfold. We don't
      // want to re-render mid-animation!
      //
      // Also, WhimsicalInstaller takes its bounding-box measurements when
      // `runInstaller` becomes true, so we want it to be in its final position
      // when this happens. Otherwise, clicking files in the air instantly
      // "teleports" them a couple inches from the mouse :/
      this.timeoutId = window.setTimeout(() => {
        this.buildProject();

        this.timeoutId = window.setTimeout(() => {
          this.setState({ runInstaller: true });
        }, 500);
      }, 600);
    }
  }

  buildProject = () => {
    const { projectName, projectType, projectIcon } = this.props;

    if (!projectName || !projectType || !projectIcon) {
      console.error('Missing one of:', {
        projectName,
        projectType,
        projectIcon,
      });
      throw new Error(
        'Tried to build project with insufficient data. See console for more info'
      );
    }

    createProject(
      { projectName, projectType, projectIcon },
      this.props.projectHomePath,
      this.handleStatusUpdate,
      this.handleError,
      this.handleComplete
    );
  };

  handleStatusUpdate = (output: any) => {
    // HACK: So, I need some way of translating the raw output from CRA
    // (and any other updates) to the `step` enums in this component.
    // It feels really gross to parse the strings in search of terms...
    // but I don't have any better ideas.
    const message = output.toString();

    if (message.match(/Installing packages/i)) {
      this.setState({
        currentBuildStep: BUILD_STEP_KEYS[2],
      });
    } else if (message.match(/Dependencies installed/i)) {
      this.setState({
        currentBuildStep: BUILD_STEP_KEYS[3],
      });
    }
  };

  handleError = (error: any) => {
    const message = error.toString();

    if (message.match(/npx: installed/i)) {
      // For unknown reasons, npx installation throws an error?
      // Everything appears to work though, so I'm just going to treat this
      // as a success.
      this.setState({
        currentBuildStep: BUILD_STEP_KEYS[1],
      });
    }
  };

  handleComplete = (project: ProjectInternal) => {
    this.setState({ isCompleted: true });

    window.setTimeout(() => {
      // Allow for the "Succeeded!" message to flash before closing the modal
      this.props.handleCompleteBuild(project);
    }, 2000);
  };

  render() {
    const { currentBuildStep, isCompleted, runInstaller } = this.state;

    return (
      <Wrapper>
        <Finished isVisible={isCompleted}>
          <FinishedInnerWrapper>
            <IconBase size={128} icon={check} />
            <Spacer size={20} />
            Project Created!
          </FinishedInnerWrapper>
        </Finished>

        <Title>Building Project...</Title>

        <WhimsicalWrapper>
          {/*
            NOTE: Hardcoding a width because the performance suffers if it's
            computed dynamically
          */}
          <WhimsicalInstaller isRunning={runInstaller} width={420} />
        </WhimsicalWrapper>

        <BuildSteps>
          {BUILD_STEP_KEYS.map(stepKey => {
            const step = BUILD_STEPS[stepKey];

            let stepStatus;

            if (isCompleted) {
              stepStatus = 'done';
            } else if (stepKey === currentBuildStep) {
              stepStatus = 'in-progress';
            } else if (
              BUILD_STEP_KEYS.indexOf(currentBuildStep) >
              BUILD_STEP_KEYS.indexOf(stepKey)
            ) {
              stepStatus = 'done';
            } else {
              stepStatus = 'upcoming';
            }

            return (
              <BuildStepProgress
                key={stepKey}
                step={step}
                status={stepStatus}
              />
            );
          })}
        </BuildSteps>
      </Wrapper>
    );
  }
}

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  align-items: center;
  height: 100%;
  background-image: linear-gradient(
    45deg,
    ${COLORS.blue[900]},
    ${COLORS.blue[800]}
  );
  border: 4px solid ${COLORS.white};
  color: ${COLORS.white};
  box-shadow: 0px 6px 60px rgba(0, 0, 0, 0.1), 0px 2px 8px rgba(0, 0, 0, 0.05);
  border-radius: 0;
  user-select: none;
`;

const BuildSteps = styled.div`
  width: 270px;
  padding-bottom: 30px;
`;

const Title = styled.h1`
  padding: 0 40px;
  font-size: 36px;
  text-align: center;
  pointer-events: none;
`;

const WhimsicalWrapper = styled.div`
  position: relative;
  width: 100%;
  /* Make sure it sits below the "Finished" overlay, when completed */
  z-index: 1;
`;

const Finished = styled.div`
  position: absolute;
  z-index: 2;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-around;
  border: 4px solid ${COLORS.white};
  background-image: linear-gradient(
    45deg,
    ${COLORS.teal[500]},
    ${COLORS.lightGreen[700]}
  );
  font-size: 32px;
  font-weight: 600;
  -webkit-font-smoothing: antialiased;
  will-change: clip-path;
  clip-path: polygon(
    ${props =>
      props.isVisible
        ? '-100% 0%, 100% 200%, 100% 0%'
        : '100% 0%, 100% 0%, 100% 0%'}
  );
  transition: clip-path 1000ms 350ms;
`;

const FinishedInnerWrapper = styled.div`
  text-align: center;
`;

export default BuildPane;
