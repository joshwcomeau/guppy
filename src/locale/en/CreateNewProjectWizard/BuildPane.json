{
  "createNewProjectWizard.buildPane.step.installingCliTool": "Installing build tool",
  "createNewProjectWizard.buildPane.step.creatingProjectDirectory": "Creating project directory",
  "createNewProjectWizard.buildPane.step.installingDependencies": "Installing dependencies",
  "createNewProjectWizard.buildPane.step.guppification": "Persisting project info",
  "createNewProjectWizard.buildPane.step.takeAWhile": "This step can take a while...",
  "createNewProjectWizard.buildPane.main.created": "Project Created!",
  "createNewProjectWizard.buildPane.main.building": "Building Project...",
  "createNewProjectWizard.buildPane.error.insufficientData": "Tried to build project with insufficient data. See console for more info"
}
