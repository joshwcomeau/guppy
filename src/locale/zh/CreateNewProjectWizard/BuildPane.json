{
  "createNewProjectWizard.buildPane.step.installingCliTool": "安装构建工具",
  "createNewProjectWizard.buildPane.step.creatingProjectDirectory": "创建项目目录",
  "createNewProjectWizard.buildPane.step.installingDependencies": "安装依赖",
  "createNewProjectWizard.buildPane.step.guppification": "项目信息持久化",
  "createNewProjectWizard.buildPane.step.takeAWhile": "您可能需要稍等一会...",
  "createNewProjectWizard.buildPane.main.created": "项目创建成功！",
  "createNewProjectWizard.buildPane.main.building": "构建项目...",
  "createNewProjectWizard.buildPane.error.insufficientData": "构建项目所需数据不足。更多信息请查看控制台"
}
