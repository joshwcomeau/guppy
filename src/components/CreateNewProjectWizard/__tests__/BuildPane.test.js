import React from 'react';
import { shallow } from 'enzyme';

import BuildPane, { BUILD_STEP_KEYS } from '../BuildPane';
import createProject from '../../../services/create-project.service';

jest.mock('../../../services/create-project.service');

describe('BuildPane component', () => {
  let wrapper;
  let instance;
  const newProject = {
    projectName: 'New project',
    projectType: 'create-react-app',
    projectIcon: 'icon',
    projectStarter: '',
  };

  const mockHandleCompleteBuild = jest.fn();

  jest.useFakeTimers();

  const shallowRenderBuildPane = project =>
    shallow(
      <BuildPane
        {...project}
        status="filling-in-form"
        handleCompleteBuild={mockHandleCompleteBuild}
      />
    );

  const testOutputs = ['Installing packages', 'Dependencies installed'];

  beforeEach(() => {
    wrapper = shallowRenderBuildPane(newProject);
    instance = wrapper.instance();

    // Mock console errors so they don't output while running the test
    jest.spyOn(console, 'error');
    console.error.mockImplementation(() => {});
  });

  afterEach(() => {
    console.error.mockRestore();
  });

  BUILD_STEP_KEYS.forEach(buildStep => {
    it(`should render build step ${buildStep}`, () => {
      wrapper.setProps({
        currentBuildStep: buildStep,
      });
      expect(wrapper).toMatchSnapshot();
    });
  });

  it('should call createProject', () => {
    wrapper.setProps({
      status: 'building-project',
    });
    jest.runAllTimers();
    expect(createProject).toHaveBeenCalled();
  });

  it('should throw error if a project prop is missing', () => {
    wrapper = shallowRenderBuildPane();
    instance = wrapper.instance();
    expect(instance.buildProject).toThrow(/insufficient data/i);
  });

  it('should handle status updated', () => {
    // Note: Instance.handleStatusUpdate is called from createProject service
    //       with data from stdout we're testing this here with a mock array of output strings
    testOutputs.forEach(output => {
      instance.handleStatusUpdate(output);
      expect(instance.state.currentBuildStep).toMatchSnapshot();
    });
  });

  it('should call handleComplete', () => {
    // Note: HandleComplete is also called from createProject service
    //       so it's OK to call this here directly as it's triggered
    //       once the service finished the work.
    instance.handleComplete(newProject);
    jest.runAllTimers();
    expect(mockHandleCompleteBuild).toHaveBeenCalledWith(newProject);
  });

  it('should handle error message', () => {
    instance.handleError('npx: installed');
    expect(instance.state.currentBuildStep).toEqual(BUILD_STEP_KEYS[1]);
  });
});
