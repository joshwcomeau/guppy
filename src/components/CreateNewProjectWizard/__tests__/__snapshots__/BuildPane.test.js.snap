// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuildPane component should handle status updated 1`] = `"installingDependencies"`;

exports[`BuildPane component should handle status updated 2`] = `"guppification"`;

exports[`BuildPane component should render build step creatingProjectDirectory 1`] = `
<styled.div>
  <styled.div
    isVisible={false}
  >
    <styled.div>
      <Icon
        fill="currentColor"
        icon={
          Object {
            "attribs": Object {
              "fill": "none",
              "stroke": "currentColor",
              "stroke-linecap": "round",
              "stroke-linejoin": "round",
              "stroke-width": "2",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "points": "20 6 9 17 4 12",
                },
                "children": Array [],
                "name": "polyline",
              },
            ],
            "viewBox": "0 0 24 24",
          }
        }
        size={128}
      />
      <styled.div
        size={20}
      />
      Project Created!
    </styled.div>
  </styled.div>
  <styled.h1>
    Building Project...
  </styled.h1>
  <styled.div>
    <WhimsicalInstaller
      isRunning={false}
      width={420}
    />
  </styled.div>
  <styled.div>
    <BuildStepProgress
      key="installingCliTool"
      status="upcoming"
      step={
        Object {
          "copy": "Installing build tool",
        }
      }
    />
    <BuildStepProgress
      key="creatingProjectDirectory"
      status="upcoming"
      step={
        Object {
          "copy": "Creating project directory",
        }
      }
    />
    <BuildStepProgress
      key="installingDependencies"
      status="upcoming"
      step={
        Object {
          "additionalCopy": "This step can take a while...",
          "copy": "Installing dependencies",
        }
      }
    />
    <BuildStepProgress
      key="guppification"
      status="upcoming"
      step={
        Object {
          "copy": "Persisting project info",
        }
      }
    />
  </styled.div>
</styled.div>
`;

exports[`BuildPane component should render build step guppification 1`] = `
<styled.div>
  <styled.div
    isVisible={false}
  >
    <styled.div>
      <Icon
        fill="currentColor"
        icon={
          Object {
            "attribs": Object {
              "fill": "none",
              "stroke": "currentColor",
              "stroke-linecap": "round",
              "stroke-linejoin": "round",
              "stroke-width": "2",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "points": "20 6 9 17 4 12",
                },
                "children": Array [],
                "name": "polyline",
              },
            ],
            "viewBox": "0 0 24 24",
          }
        }
        size={128}
      />
      <styled.div
        size={20}
      />
      Project Created!
    </styled.div>
  </styled.div>
  <styled.h1>
    Building Project...
  </styled.h1>
  <styled.div>
    <WhimsicalInstaller
      isRunning={false}
      width={420}
    />
  </styled.div>
  <styled.div>
    <BuildStepProgress
      key="installingCliTool"
      status="upcoming"
      step={
        Object {
          "copy": "Installing build tool",
        }
      }
    />
    <BuildStepProgress
      key="creatingProjectDirectory"
      status="upcoming"
      step={
        Object {
          "copy": "Creating project directory",
        }
      }
    />
    <BuildStepProgress
      key="installingDependencies"
      status="upcoming"
      step={
        Object {
          "additionalCopy": "This step can take a while...",
          "copy": "Installing dependencies",
        }
      }
    />
    <BuildStepProgress
      key="guppification"
      status="upcoming"
      step={
        Object {
          "copy": "Persisting project info",
        }
      }
    />
  </styled.div>
</styled.div>
`;

exports[`BuildPane component should render build step installingCliTool 1`] = `
<styled.div>
  <styled.div
    isVisible={false}
  >
    <styled.div>
      <Icon
        fill="currentColor"
        icon={
          Object {
            "attribs": Object {
              "fill": "none",
              "stroke": "currentColor",
              "stroke-linecap": "round",
              "stroke-linejoin": "round",
              "stroke-width": "2",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "points": "20 6 9 17 4 12",
                },
                "children": Array [],
                "name": "polyline",
              },
            ],
            "viewBox": "0 0 24 24",
          }
        }
        size={128}
      />
      <styled.div
        size={20}
      />
      Project Created!
    </styled.div>
  </styled.div>
  <styled.h1>
    Building Project...
  </styled.h1>
  <styled.div>
    <WhimsicalInstaller
      isRunning={false}
      width={420}
    />
  </styled.div>
  <styled.div>
    <BuildStepProgress
      key="installingCliTool"
      status="upcoming"
      step={
        Object {
          "copy": "Installing build tool",
        }
      }
    />
    <BuildStepProgress
      key="creatingProjectDirectory"
      status="upcoming"
      step={
        Object {
          "copy": "Creating project directory",
        }
      }
    />
    <BuildStepProgress
      key="installingDependencies"
      status="upcoming"
      step={
        Object {
          "additionalCopy": "This step can take a while...",
          "copy": "Installing dependencies",
        }
      }
    />
    <BuildStepProgress
      key="guppification"
      status="upcoming"
      step={
        Object {
          "copy": "Persisting project info",
        }
      }
    />
  </styled.div>
</styled.div>
`;

exports[`BuildPane component should render build step installingDependencies 1`] = `
<styled.div>
  <styled.div
    isVisible={false}
  >
    <styled.div>
      <Icon
        fill="currentColor"
        icon={
          Object {
            "attribs": Object {
              "fill": "none",
              "stroke": "currentColor",
              "stroke-linecap": "round",
              "stroke-linejoin": "round",
              "stroke-width": "2",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "points": "20 6 9 17 4 12",
                },
                "children": Array [],
                "name": "polyline",
              },
            ],
            "viewBox": "0 0 24 24",
          }
        }
        size={128}
      />
      <styled.div
        size={20}
      />
      Project Created!
    </styled.div>
  </styled.div>
  <styled.h1>
    Building Project...
  </styled.h1>
  <styled.div>
    <WhimsicalInstaller
      isRunning={false}
      width={420}
    />
  </styled.div>
  <styled.div>
    <BuildStepProgress
      key="installingCliTool"
      status="upcoming"
      step={
        Object {
          "copy": "Installing build tool",
        }
      }
    />
    <BuildStepProgress
      key="creatingProjectDirectory"
      status="upcoming"
      step={
        Object {
          "copy": "Creating project directory",
        }
      }
    />
    <BuildStepProgress
      key="installingDependencies"
      status="upcoming"
      step={
        Object {
          "additionalCopy": "This step can take a while...",
          "copy": "Installing dependencies",
        }
      }
    />
    <BuildStepProgress
      key="guppification"
      status="upcoming"
      step={
        Object {
          "copy": "Persisting project info",
        }
      }
    />
  </styled.div>
</styled.div>
`;
